File: app/src/main/java/com/example/ui/viewmodel/CumaViewModel.kt
package com.example.ui.viewmodel

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.CumaApp
import com.example.data.model.FridayMessage
import com.example.data.model.FridaySchedule
import com.example.data.model.RecipientContact
import com.example.data.model.SentLog
import com.example.data.sample.DefaultData
import com.example.util.AlarmScheduler
import com.example.util.CardGenerator
import com.example.util.ContactsHelper
import com.example.util.PinterestHelper
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

data class SequentialSendStatus(
    val isActive: Boolean = false,
    val contacts: List<RecipientContact> = emptyList(),
    val currentIndex: Int = 0,
    val currentImageUri: Uri? = null,
    val currentMessage: String = ""
)

class CumaViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = (application as CumaApp).repository

    val recipients: StateFlow<List<RecipientContact>> = repository.getAllRecipients()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val selectedRecipients: StateFlow<List<RecipientContact>> = repository.getSelectedRecipients()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val messages: StateFlow<List<FridayMessage>> = repository.getAllMessages()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val schedule: StateFlow<FridaySchedule?> = repository.getScheduleFlow()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), DefaultData.defaultSchedule)

    val sentLogs: StateFlow<List<SentLog>> = repository.getSentLogs()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _selectedCategory = MutableStateFlow("Tümü")
    val selectedCategory: StateFlow<String> = _selectedCategory.asStateFlow()

    private val _recipientFilterGroup = MutableStateFlow("Tümü")
    val recipientFilterGroup: StateFlow<String> = _recipientFilterGroup.asStateFlow()

    private val _previewCardUri = MutableStateFlow<Uri?>(null)
    val previewCardUri: StateFlow<Uri?> = _previewCardUri.asStateFlow()

    private val _isProcessing = MutableStateFlow(false)
    val isProcessing: StateFlow<Boolean> = _isProcessing.asStateFlow()

    private val _statusMessage = MutableStateFlow<String?>(null)
    val statusMessage: StateFlow<String?> = _statusMessage.asStateFlow()

    // Sequential sender state
    private val _sequentialSendState = MutableStateFlow(SequentialSendStatus())
    val sequentialSendState: StateFlow<SequentialSendStatus> = _sequentialSendState.asStateFlow()

    fun setSelectedCategory(category: String) {
        _selectedCategory.value = category
    }

    fun setRecipientFilterGroup(group: String) {
        _recipientFilterGroup.value = group
    }

    fun clearStatusMessage() {
        _statusMessage.value = null
    }

    // Recipient Actions
    fun toggleRecipient(id: Long, isSelected: Boolean) {
        viewModelScope.launch {
            repository.toggleRecipientSelection(id, isSelected)
        }
    }

    fun selectAllRecipients(select: Boolean) {
        viewModelScope.launch {
            repository.selectAllRecipients(select)
        }
    }

    fun addRecipient(name: String, phoneNumber: String, groupName: String = "Aile") {
        viewModelScope.launch {
            if (name.isNotBlank() && phoneNumber.isNotBlank()) {
                repository.addRecipient(
                    RecipientContact(
                        name = name.trim(),
                        phoneNumber = phoneNumber.trim(),
                        groupName = groupName,
                        isSelected = true
                    )
                )
                _statusMessage.value = "$name rehbere eklendi."
            }
        }
    }

    fun deleteRecipient(recipient: RecipientContact) {
        viewModelScope.launch {
            repository.deleteRecipient(recipient)
        }
    }

    fun importPhoneContacts() {
        viewModelScope.launch {
            _isProcessing.value = true
            try {
                val contacts = ContactsHelper.fetchPhoneContacts(getApplication())
                if (contacts.isNotEmpty()) {
                    repository.addRecipients(contacts)
                    _statusMessage.value = "${contacts.size} kişi telefon rehberinden aktarıldı."
                } else {
                    _statusMessage.value = "Rehberde kişi bulunamadı veya izin verilmedi."
                }
            } catch (e: Exception) {
                _statusMessage.value = "Kişiler alınamadı: ${e.message}"
            } finally {
                _isProcessing.value = false
            }
        }
    }

    // Message Actions
    fun addMessage(title: String, content: String, category: String = "Özel") {
        viewModelScope.launch {
            if (content.isNotBlank()) {
                val id = repository.addMessage(
                    FridayMessage(
                        title = title.ifBlank { "Özel Mesaj" },
                        content = content.trim(),
                        category = category,
                        isCustom = true
                    )
                )
                updateSelectedMessage(id)
                _statusMessage.value = "Yeni Cuma mesajı kaydedildi ve seçildi."
            }
        }
    }

    fun selectRandomMessage() {
        viewModelScope.launch {
            val list = messages.value
            if (list.isNotEmpty()) {
                val random = list.random()
                updateSelectedMessage(random.id)
                _statusMessage.value = "Bu Cuma için yeni bir mesaj seçildi: ${random.title}"
            }
        }
    }

    fun toggleMessageFavorite(id: Long, isFavorite: Boolean) {
        viewModelScope.launch {
            repository.toggleMessageFavorite(id, isFavorite)
        }
    }

    fun deleteMessage(message: FridayMessage) {
        viewModelScope.launch {
            repository.deleteMessage(message)
        }
    }

    // Schedule & Visual Typography Actions
    fun updateScheduleTime(hour: Int, minute: Int) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            val updated = curr.copy(hour = hour, minute = minute)
            repository.updateSchedule(updated)
            if (updated.isEnabled) {
                AlarmScheduler.scheduleFridayReminder(getApplication(), hour, minute)
            }
            _statusMessage.value = "Zamanlayıcı güncellendi: ${AlarmScheduler.formatTime(hour, minute)}"
        }
    }

    fun toggleScheduleEnabled(enabled: Boolean) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            val updated = curr.copy(isEnabled = enabled)
            repository.updateSchedule(updated)
            if (enabled) {
                AlarmScheduler.scheduleFridayReminder(getApplication(), updated.hour, updated.minute)
                _statusMessage.value = "Cuma zamanlayıcısı aktif edildi."
            } else {
                AlarmScheduler.cancelReminder(getApplication())
                _statusMessage.value = "Cuma zamanlayıcısı duraklatıldı."
            }
        }
    }

    fun updateSelectedMessage(messageId: Long) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            repository.updateSchedule(curr.copy(selectedMessageId = messageId))
            generatePreviewCard()
        }
    }

    fun updateSelectedImageKey(imageKey: String) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            repository.updateSchedule(curr.copy(selectedImageKey = imageKey))
            generatePreviewCard()
        }
    }

    fun pickCustomPhotoUri(uri: Uri) {
        viewModelScope.launch {
            updateSelectedImageKey(uri.toString())
            _statusMessage.value = "Fotoğraf galerinizden başarıyla seçildi!"
        }
    }

    fun updateFontStyle(style: String) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            repository.updateSchedule(curr.copy(fontStyle = style))
            generatePreviewCard()
        }
    }

    fun updateFrameStyle(style: String) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            repository.updateSchedule(curr.copy(frameStyle = style))
            generatePreviewCard()
        }
    }

    fun updateTextSizeModifier(modifier: Float) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            repository.updateSchedule(curr.copy(textSizeModifier = modifier))
            generatePreviewCard()
        }
    }

    fun updateTextColorHex(hex: String) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            repository.updateSchedule(curr.copy(textColorHex = hex))
            generatePreviewCard()
        }
    }

    fun updateSenderSignature(signature: String) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            repository.updateSchedule(curr.copy(senderSignature = signature))
            generatePreviewCard()
        }
    }

    fun toggleMergeTextOnImage(merge: Boolean) {
        viewModelScope.launch {
            val curr = schedule.value ?: DefaultData.defaultSchedule
            repository.updateSchedule(curr.copy(mergeTextOnImage = merge))
        }
    }

    // Card Generation
    fun generatePreviewCard() {
        viewModelScope.launch {
            val sched = schedule.value ?: DefaultData.defaultSchedule
            val msg = messages.value.find { it.id == sched.selectedMessageId }
                ?: messages.value.firstOrNull()
                ?: return@launch

            val uri = CardGenerator.generateCardImage(
                context = getApplication(),
                imageKeyOrPath = sched.selectedImageKey,
                messageText = msg.content,
                signature = sched.senderSignature,
                recipientName = "",
                fontStyle = sched.fontStyle,
                frameStyle = sched.frameStyle,
                textSizeModifier = sched.textSizeModifier,
                textColorHex = sched.textColorHex
            )
            _previewCardUri.value = uri
        }
    }

    // Download Pinterest / Web Image
    fun downloadWebImage(url: String, onComplete: (Boolean) -> Unit) {
        viewModelScope.launch {
            _isProcessing.value = true
            val localPath = PinterestHelper.downloadImageFromUrl(getApplication(), url)
            _isProcessing.value = false
            if (localPath != null) {
                updateSelectedImageKey(localPath)
                _statusMessage.value = "Pinterest resmi başarıyla indirildi ve seçildi!"
                onComplete(true)
            } else {
                _statusMessage.value = "Resim indirilemedi. Lütfen geçerli bir Pinterest veya resim linki girin."
                onComplete(false)
            }
        }
    }

    // Sequential Queue Dispatcher
    fun startSequentialSend() {
        viewModelScope.launch {
            val targets = selectedRecipients.value
            if (targets.isEmpty()) {
                _statusMessage.value = "Lütfen önce mesaj gönderilecek kişileri seçin veya rehberden aktarın."
                return@launch
            }

            val sched = schedule.value ?: DefaultData.defaultSchedule
            val msg = messages.value.find { it.id == sched.selectedMessageId }
                ?: messages.value.firstOrNull()

            val rawMessage = msg?.content ?: "Hayırlı Cumalar dilerim."

            val cardUri = if (sched.mergeTextOnImage) {
                CardGenerator.generateCardImage(
                    context = getApplication(),
                    imageKeyOrPath = sched.selectedImageKey,
                    messageText = rawMessage,
                    signature = sched.senderSignature,
                    fontStyle = sched.fontStyle,
                    frameStyle = sched.frameStyle,
                    textSizeModifier = sched.textSizeModifier,
                    textColorHex = sched.textColorHex
                )
            } else null

            _sequentialSendState.value = SequentialSendStatus(
                isActive = true,
                contacts = targets,
                currentIndex = 0,
                currentImageUri = cardUri,
                currentMessage = rawMessage
            )
        }
    }

    fun advanceSequentialSend(isSent: Boolean = true) {
        val currState = _sequentialSendState.value
        if (!currState.isActive) return

        val currentContact = currState.contacts.getOrNull(currState.currentIndex)
        if (currentContact != null && isSent) {
            viewModelScope.launch {
                repository.markContactSent(currentContact.id)
                repository.logSent(
                    contactName = currentContact.name,
                    phoneNumber = currentContact.phoneNumber,
                    messageSnippet = currState.currentMessage
                )
            }
        }

        val nextIndex = currState.currentIndex + 1
        if (nextIndex < currState.contacts.size) {
            _sequentialSendState.value = currState.copy(currentIndex = nextIndex)
        } else {
            // Completed!
            _sequentialSendState.value = SequentialSendStatus(isActive = false)
            _statusMessage.value = "Tüm Cuma mesajları başarıyla iletildi! Allah kabul etsin. 🤲"
        }
    }

    fun cancelSequentialSend() {
        _sequentialSendState.value = SequentialSendStatus(isActive = false)
    }
}

File: iosApp/ViewModels/CumaViewModel.swift
import Foundation
import SwiftUI
import Combine

final class CumaViewModel: ObservableObject {
    // MARK: - Published State
    @Published var recipients: [RecipientContact] = [] // Starts EMPTY as requested!
    @Published var messages: [FridayMessage] = []
    @Published var schedule: FridaySchedule = FridaySchedule()
    @Published var sentLogs: [SentLog] = []

    @Published var previewCardImage: UIImage? = nil
    @Published var customPickedImage: UIImage? = nil
    @Published var statusToast: String? = nil

    // Sequential Queue State
    @Published var isQueueActive: Bool = false
    @Published var currentQueueIndex: Int = 0

    init() {
        loadInitialMessages()
        generatePreviewCard()
    }

    var selectedRecipients: [RecipientContact] {
        recipients.filter { $0.isSelected }
    }

    var currentMessage: FridayMessage? {
        messages.first(where: { $0.id == schedule.selectedMessageId }) ?? messages.first
    }

    // MARK: - Contacts Management
    func importPhoneContacts() {
        ContactsManager.shared.requestAccess { [weak self] granted in
            guard let self = self else { return }
            if granted {
                let fetched = ContactsManager.shared.fetchContacts()
                self.recipients = fetched
                self.statusToast = "\(fetched.count) kişi rehberden aktarıldı."
            } else {
                self.statusToast = "Rehber erişimine izin verilmedi."
            }
        }
    }

    func addRecipient(name: String, phoneNumber: String, groupName: String = "Aile") {
        let newContact = RecipientContact(name: name, phoneNumber: phoneNumber, groupName: groupName, isSelected: true)
        recipients.append(newContact)
        statusToast = "\(name) listeye eklendi."
    }

    func toggleRecipientSelection(id: UUID) {
        if let index = recipients.firstIndex(where: { $0.id == id }) {
            recipients[index].isSelected.toggle()
        }
    }

    func selectAllRecipients(_ select: Bool) {
        for i in 0..<recipients.count {
            recipients[i].isSelected = select
        }
    }

    func deleteRecipient(at offsets: IndexSet) {
        recipients.remove(atOffsets: offsets)
    }

    // MARK: - Messages Management
    func selectRandomMessage() {
        if let random = messages.randomElement() {
            schedule.selectedMessageId = random.id
            generatePreviewCard()
            statusToast = "Rastgele Cuma mesajı seçildi: \(random.title)"
        }
    }

    func toggleMessageFavorite(id: Int) {
        if let idx = messages.firstIndex(where: { $0.id == id }) {
            messages[idx].isFavorite.toggle()
        }
    }

    func addCustomMessage(title: String, content: String) {
        let newId = (messages.map(\.id).max() ?? 0) + 1
        let msg = FridayMessage(id: newId, title: title, content: content, category: "Özel", isFavorite: true, isCustom: true)
        messages.insert(msg, at: 0)
        schedule.selectedMessageId = newId
        generatePreviewCard()
        statusToast = "Yeni Cuma mesajınız eklendi ve seçildi."
    }

    // MARK: - Schedule & Card Customization
    func updateScheduleTime(hour: Int, minute: Int) {
        schedule.hour = hour
        schedule.minute = minute
        if schedule.isEnabled {
            NotificationManager.shared.scheduleFridayReminder(hour: hour, minute: minute)
        }
        statusToast = "Zamanlayıcı güncellendi: \(String(format: "%02d:%02d", hour, minute))"
    }

    func toggleScheduleEnabled(_ enabled: Bool) {
        schedule.isEnabled = enabled
        if enabled {
            NotificationManager.shared.scheduleFridayReminder(hour: schedule.hour, minute: schedule.minute)
            statusToast = "Cuma hatırlatıcısı aktif."
        } else {
            NotificationManager.shared.cancelReminder()
            statusToast = "Cuma hatırlatıcısı duraklatıldı."
        }
    }

    func setCustomImage(_ image: UIImage) {
        customPickedImage = image
        schedule.selectedImageKey = "custom_user_photo"
        generatePreviewCard()
        statusToast = "Fotoğraf başarıyla güncellendi!"
    }

    func setSelectedImageKey(_ key: String) {
        schedule.selectedImageKey = key
        customPickedImage = nil
        generatePreviewCard()
    }

    func generatePreviewCard() {
        let baseImage: UIImage
        if let custom = customPickedImage, schedule.selectedImageKey == "custom_user_photo" {
            baseImage = custom
        } else {
            // Default placeholder spiritual mosque representation if asset isn't present
            baseImage = UIImage(named: schedule.selectedImageKey) ?? createPlaceholderMosqueImage()
        }

        let msgText = currentMessage?.content ?? "Hayırlı Cumalar dilerim."

        previewCardImage = CardGenerator.shared.generateCardImage(
            backgroundImage: baseImage,
            messageText: msgText,
            signature: schedule.senderSignature,
            recipientName: "",
            fontStyle: schedule.fontStyle,
            frameStyle: schedule.frameStyle,
            textSizeModifier: schedule.textSizeModifier,
            textColorHex: schedule.textColorHex
        )
    }

    private func createPlaceholderMosqueImage() -> UIImage {
        let size = CGSize(width: 800, height: 1000)
        UIGraphicsBeginImageContext(size)
        let ctx = UIGraphicsGetCurrentContext()
        ctx?.setFillColor(UIColor(red: 0, green: 0.28, blue: 0.20, alpha: 1.0).cgColor)
        ctx?.fill(CGRect(origin: .zero, size: size))
        let img = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return img ?? UIImage()
    }

    // MARK: - Queue Dispatcher
    func startSequentialQueue() {
        guard !selectedRecipients.isEmpty else {
            statusToast = "Lütfen önce mesaj gönderilecek kişileri seçin."
            return
        }
        currentQueueIndex = 0
        isQueueActive = true
    }

    func advanceQueue(didSend: Bool) {
        if didSend, let contact = selectedRecipients[safe: currentQueueIndex] {
            sentLogs.insert(SentLog(contactName: contact.name, phoneNumber: contact.phoneNumber, messageSnippet: currentMessage?.content ?? ""), at: 0)
        }

        if currentQueueIndex + 1 < selectedRecipients.count {
            currentQueueIndex += 1
        } else {
            isQueueActive = false
            statusToast = "Tüm Cuma mesajları gönderildi! 🤲"
        }
    }

    func cancelQueue() {
        isQueueActive = false
    }

    // MARK: - Initial Message Library
    private func loadInitialMessages() {
        messages = [
            FridayMessage(id: 1, title: "Cuma Suresi 9. Ayet", content: "“Ey iman edenler! Cuma günü namaza çağrıldığınız zaman hemen Allah'ı anmaya koşun ve alışverişi bırakın. Eğer bilirseniz bu sizin için daha hayırlıdır.” (Cuma Suresi, 9)\n\nHayırlı ve bereketli Cumalar dilerim.", category: "Ayet & Hadis", isFavorite: true),
            FridayMessage(id: 2, title: "En Hayırlı Gün", content: "Resulullah (s.a.v.) buyurdu: “Üzerine güneşin doğduğu en hayırlı gün Cuma günüdür.”\n\nRabbim bu mübarek günün hürmetine gönlünüzdeki hayırlı duaları kabul eylesin. Hayırlı Cumalar {isim}!", category: "Ayet & Hadis", isFavorite: true),
            FridayMessage(id: 3, title: "İcabet Saati Müjdesi", content: "“Cuma gününde bir saat vardır ki, Müslüman bir kul o saatte Allah'tan bir hayır dilerse, Allah ona mutlaka verir.” (Buhârî)\n\nDualarımızın makbul olduğu hayırlı, nurlu Cumalar.", category: "Ayet & Hadis", isFavorite: true),
            FridayMessage(id: 4, title: "Huzur ve Afiyet Duası", content: "Allah'ım! Kalbimize ferahlık, hanemize huzur ve bereket, bedenimize sağlık, ömrümüze afiyet ihsan eyle. Bizleri sevdiklerimizle cennette buluştur.\n\nMübarek Cuma gününüz hayırlara vesile olsun.", category: "Dualar & Bereket", isFavorite: true),
            FridayMessage(id: 5, title: "Gönül Şifası ve Genişlik", content: "Rabbim, ömrümüzü bereketli, amellerimizi ihlaslı, dualarımızı kabul eylesin. Dertlerimize deva, gönüllerimize şifa, hanelerimize neşe nasip etsin.\n\nHayırlı ve nurlu Cumalar dilerim {isim}.", category: "Dualar & Bereket", isFavorite: true),
            FridayMessage(id: 6, title: "Samimi ve Nurlu", content: "Gönlünüzden geçen her hayırlı duanın kabul olması dileğiyle... Hayırlı, huzurlu ve bereketli Cumalar!", category: "Samimi & Kısa", isFavorite: true),
            FridayMessage(id: 7, title: "Büyüklerimize Hürmet", content: "Hürmet ve sevgilerimle ellerinizden öperim. Rabbim başımızdan eksik etmesin, sağlıklı ve hayırlı ömürler versin.\n\nMübarek Cuma gününüz kutlu olsun {isim}.", category: "Akraba & Büyükler", isFavorite: true),
            FridayMessage(id: 8, title: "Rabbena Duası", content: "“Rabbimiz! Bize dünyada da iyilik ver, ahirette de iyilik ver ve bizi ateş azabından koru.” (Bakara, 201)\n\nDualarımızın kabul olması dileğiyle hayırlı Cumalar.", category: "Peygamberimizin Duaları", isFavorite: true)
        ]
    }
}

extension Collection {
    subscript(safe index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
